### PR TITLE
Redesign database unlock widget

### PR DESCRIPTION
**Pull request by phoerious**
## Description
With this change we get rid of the confusing key component checkboxes. Now a component is either there or not (if left empty). There is no redundant distinction between "unset" and "emtpy" anymore.
For compatibility with older databases that have "empty" passwords, KeePassXC will ask if the user wants to retry with an empty password if unlocking failed and the password field was left blank.

Besides these functional changes, the widget's layout has been rearranged to be more compact, less stretched out (e.g. input fields do not fill the full window width anymore), and more user-friendly
by providing a help tooltip for the hardware key field and accessible descriptions for screen readers.

**Note:** I am not totally satisfied with the layout yet. For instance, I'd like to make the background brighter so that the unlock panel stands out more. However, I am not sure how to do that without breaking all sorts of themes (particularly dark themes).

(The TouchID checkbox is only in the screenshot so you can see how it looks, it's not visible in the committed source code.)

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Refactor (significant modification to existing code)
- ✅ New feature (non-breaking change which adds functionality)

## Screenshots
[TIP]:  # ( Do not include screenshots of your actual database! )
image

image

image

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )
Tests were adjusted to accommodate the changes.

## Checklist:
[NOTE]: # ( Please go over all the following points. )
[NOTE]: # ( Again, remove any lines which don't apply. )
[NOTE]: # ( Pull Requests that don't fulfill all [REQUIRED] requisites are likely )
[NOTE]: # ( to be sent back to you for correction or will be rejected.  )
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

### CODE REVIEW

File: src/gui/DatabaseOpenWidget.cpp [hunk 1]
```diff
@@ -46,18 +46,12 @@ DatabaseOpenWidget::DatabaseOpenWidget(QWidget* parent)
 
     m_ui->messageWidget->setHidden(true);
 
-    QFont font = m_ui->labelHeadline->font();
-    font.setBold(true);
-    font.setPointSize(font.pointSize() + 2);
-    m_ui->labelHeadline->setFont(font);
+    m_ui->labelHeadline->setText(tr("Unlock KeePassXC Database"));
```
Comment (droidmonkey) [inline src/gui/DatabaseOpenWidget.cpp]: I find the font to be rather small and uninviting. We should place the purpose of the dialog (this text) and the database we are unlocking in a prominent and dedicated position. Here is a small mockup:

image

File: src/gui/DatabaseOpenWidget.cpp [hunk 2]
```diff
@@ -137,13 +129,12 @@ void DatabaseOpenWidget::load(const QString& filename)
 {
     m_filename = filename;
 
-    m_ui->labelFilename->setText(filename);
-
+    m_ui->comboKeyFile->addItem(tr("None"), -1);
```
Comment (droidmonkey) [inline src/gui/DatabaseOpenWidget.cpp]: To align with the hardware key, recommend "Select file..."

File: src/gui/DatabaseOpenWidget.cpp [hunk 3]
```diff
@@ -194,6 +183,26 @@ void DatabaseOpenWidget::openDatabase()
...
     QApplication::restoreOverrideCursor();
     if (!ok) {
+        if (m_ui->editPassword->text().isEmpty() && !m_retryUnlockWithEmptyPassword) {
+            QScopedPointer<QMessageBox> msgBox(new QMessageBox(this));
+            msgBox->setIcon(QMessageBox::Critical);
+            msgBox->setWindowTitle(tr("Unlock failed and no password given"));
+            msgBox->setText(tr("Unlocking the database failed and you did not enter a password.\n"
+                               "Do you want to retry with an \"empty\" password instead?\n\n"
+                               "If your database really has no password and this keeps happening,\n"
+                               "go to \"Database Settings / Security\" and reset your password there."));
```
Comment (droidmonkey) [inline src/gui/DatabaseOpenWidget.cpp]: Rephrase this to:
> To prevent this error from appearing you must go to "Database Settings / Security" and reset your password.

File: src/gui/DatabaseOpenWidget.ui [hunk 4]
```diff
@@ -31,27 +31,308 @@
...
+                         </property>
+                         <property name="text">
+                          <string>Refresh</string>
+                         </property>
+                        </widget>
+                       </item>
+                       <item row="0" column="0">
+                        <widget class="QLabel" name="label_4">
+                         <property name="text">
+                          <string>YubiKey:</string>
```
Comment (droidmonkey) [inline src/gui/DatabaseOpenWidget.ui]: Recommend naming this "Hardware Token" and add a help tooltip that shows the following when hovered:

> KeePassXC supports Yubikey and OnlyKey hardware tokens with slots configured with HMAC challenge-response. To read more about this solution, please visit https://keepassxc.org/docs#hwtoken.

OnlyKey support is being added to 2.5.0 (eventually)

File: src/gui/DatabaseOpenWidget.cpp (hunk 1 above)
Comment (droidmonkey) [reply src/gui/DatabaseOpenWidget.cpp]: You can wrap the text with ```<h1>``` tags to make it bigger, but it still seems small to me on Windows 10.

Comment (phoerious) [reply src/gui/DatabaseOpenWidget.cpp]: On Windows everything looks small. I think we can put it above and make it larger, but this looks a bit unbalanced to me. Need to think about how we best solve this.

File: src/gui/DatabaseOpenWidget.cpp (hunk 3 above)
Comment (phoerious) [reply src/gui/DatabaseOpenWidget.cpp]: This error is specific for when there is no password provided. I kinda wanted to include that into the description to make it more explicit for the user.

File: src/gui/DatabaseOpenWidget.ui (hunk 4 above)
Comment (phoerious) [reply src/gui/DatabaseOpenWidget.ui]: Agree, although "Hardware Token" is quite arcane. Any easier and less technical terms? We had "Challenge Response" before, but that's even harder to understand.

File: src/gui/DatabaseOpenWidget.cpp (hunk 2 above)
Comment (phoerious) [reply src/gui/DatabaseOpenWidget.cpp]: I had that before, but wasn't sure if it's clear enough that you have to select that in order to get rid of a key file. Will do, though.

File: src/gui/DatabaseOpenWidget.cpp (hunk 3 above)
Comment (phoerious) [reply src/gui/DatabaseOpenWidget.cpp]: We could even remove the message and retry implicitly. We'd effectively hide the different between an unset and an empty password this way and the user would never notice. Only problem: unlocking takes twice as long and they YubiKey will be challenged twice.

File: src/gui/DatabaseOpenWidget.cpp (hunk 2 above)
Comment (droidmonkey) [reply src/gui/DatabaseOpenWidget.cpp]: hmmm good point. Perhaps you can add the "clear" button to the right similar to the search box. When clicked it goes back to "select file...".

File: src/gui/DatabaseOpenWidget.cpp (hunk 1 above)
Comment (droidmonkey) [reply src/gui/DatabaseOpenWidget.cpp]: Oh right, well then at least the file path would be nice.

I like the way OKThanks mocked up our "onboarding" process. This is kind of the look I was going for by titling the widget:

image

File: src/gui/DatabaseOpenWidget.cpp (hunk 3 above)
Comment (droidmonkey) [reply src/gui/DatabaseOpenWidget.cpp]: Sorry I meant just rephrase the last sentence. Keep everything else.

File: src/gui/DatabaseOpenWidget.ui (hunk 4 above)
Comment (droidmonkey) [reply src/gui/DatabaseOpenWidget.ui]: "Security Key"
"Multi-Factor Token"

File: src/gui/DatabaseOpenWidget.cpp (hunk 1 above)
Comment (phoerious) [reply src/gui/DatabaseOpenWidget.cpp]: I would like to give KeePassXC a more modern look (perhaps not as "web-appy" or "bootstrappy" as the screenshot above, but similar), but it's hard to do with Qt if we want to go for a platform-native look. I think the only way here would be QtQuick with custom styling. But I fear we would alienate a lot of users with an aggressive redesign of everything.

File: src/gui/DatabaseOpenWidget.cpp (hunk 2 above)
Comment (phoerious) [reply src/gui/DatabaseOpenWidget.cpp]: That would indeed be an option.

File: src/gui/DatabaseOpenWidget.ui (hunk 4 above)
Comment (phoerious) [reply src/gui/DatabaseOpenWidget.ui]: I think both "key" and "token" are quite ambiguous. "key" probably even more than "token", which is more of a foreign word to lay users I guess.

File: src/gui/DatabaseOpenWidget.cpp (hunk 1 above)
Comment (droidmonkey) [reply src/gui/DatabaseOpenWidget.cpp]: There is a middle ground here. We are currently on the "nothing custom, nothing special, very technical" side of the graph.

File: src/gui/DatabaseOpenWidget.ui (hunk 4 above)
Comment (droidmonkey) [reply src/gui/DatabaseOpenWidget.ui]: But if you own one you know the term (which is the point) and the help tooltip helps educate the lay user.

Something like this:

img

File: src/gui/DatabaseOpenWidget.cpp (hunk 1 above)
Comment (phoerious) [reply src/gui/DatabaseOpenWidget.cpp]: I know. I would love to do a bit more in the "styled" direction, but it's hard to do with QWidgets.

**Comment by droidmonkey** [comment]
When the window is resized down to its minimum extents most of the buttons scrunch up:

image

**Comment by phoerious** [comment]
Oh dear, that is one of these extremely hard to fix Qt layout annoyances again. We had the same thing with the alphabet selection buttons of the password generator.

**Comment by phoerious** [comment]
I set some layout constraints to MinimumSize. Now it cannot be resized further than this:

image

The space above the OK/Cancel buttons is still being squished upwards when you resize to the smallest possible window size, but I think that's actually an okay feature here.
